Chapter 8/Lesson 4/homework/homework.js

```
@@ -0,0 +1,316 @@
+//             arr[3](arr[4]); return false;
+//         }
+//         return true;
+//     }
+// }
+
+// let validateName = composeValidators(
+//     isRequired,
+//     isString,
+//     hasLengthBetween(2, 100)
+// );
+
+// isGreaterThan = composeValidators(isGreaterThan);
+
+// let validateAge = composeValidators(
+//     isNumber,
+//     isGreaterThan(0)
+// );
+
+
+// function withValidation([f1, f2]) {
+//     result = [];
+//     return function () {
+//         result[0] = f1.call(this, arguments[0]);
+//         result[1] = f2.call(this, arguments[1]);
+//         if (result[0] == false || result[1] == false) return;
+//         console.log("All parameters are valid. The function doesn't throw an error.");
+//     }
+// }
+
+// const createUserWithValidation = withValidation([
+//     validateName,
+//     validateAge
+// ])("Jon", 12);
+
+
+
+function isRequired(value) {
+    if (!value) {
+        console.error("Name is required."); return false;
+    }
+}
+
+function isString(value) {
+    if (typeof value !== "string") {
+        console.error("Name must be a string."); return false;
+    }
+}
+
+function hasLengthBetween(a, b) {
+    return function (value) {
+        if (value.length < a || value.length > b) {
+            console.error(`Name must be between ${a} and ${b} characters long.`);
+            return false;
+        }
+    }
+}
+
+
+function isNumber(value) {
+    if (typeof value !== "number") {
+        console.error("Age must be a number.");
+        return false;
+    }
+}
+
+function isGreaterThan(number) {
+    return function (value) {
+        if (value < number) {
+            console.error(`Age must be greater than ${number}`);
+            return false;
+        }
+    }
+}
+
+
+function composeValidators() {
+    let arr = [].concat(...arguments);
+    return function (value) {
+        for (let i = 0; i < arr.length; i++) {
+            let result = arr[i](value);
+            if (result == false) return false;
```

According to the task, the function should return the first error message it encounters.